modules/dnn/src/tensorflow/tf_importer.cpp

```diff
@@ -510,2051 +510,2301 @@ class TFImporter
+    void parseL2Normalize        (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
+    void parsePriorBox           (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
+    void parseSoftmax            (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
+    void parseCropAndResize      (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
+    void parseMean               (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
+    void parsePack               (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
+    void parseClipByValue        (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
+    void parseLeakyRelu          (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
+    void parseActivation         (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
+
+    void parseCustomLayer        (tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams);
 };
 
-TFImporter::TFImporter(Net& net, const char *model, const char *config)
-    : dstNet(net)
+Mutex& getDispatchMapMutex()
 {
-    if (model && model[0])
-    {
-        CV_LOG_DEBUG(NULL, "DNN/TF: processing TensorFlow model from file: " << model);
-        ReadTFNetParamsFromBinaryFileOrDie(model, &netBin);
-    }
-    if (config && config[0])
+    static Mutex* volatile instance = NULL;
+    if (instance == NULL)
     {
-        CV_LOG_DEBUG(NULL, "DNN/TF: processing TensorFlow config from file: " << config);
-        ReadTFNetParamsFromTextFileOrDie(config, &netTxt);
+        cv::AutoLock lock(getInitializationMutex());
+        if (instance == NULL)
+            instance = new Mutex();
     }
-
-    populateNet();
+    return *instance;
 }
 
-TFImporter::TFImporter(
-        Net& net,
-        const char *dataModel, size_t lenModel,
-        const char *dataConfig, size_t lenConfig
-)
-    : dstNet(net)
+const TFImporter::DispatchMap& TFImporter::getDispatchMap()
 {
-    if (dataModel != NULL && lenModel > 0)
-    {
-        CV_LOG_DEBUG(NULL, "DNN/TF: processing TensorFlow model from memory (" << lenModel << " bytes)");
-        ReadTFNetParamsFromBinaryBufferOrDie(dataModel, lenModel, &netBin);
-    }
-    if (dataConfig != NULL && lenConfig > 0)
+    static const TFImporter::DispatchMap* volatile dispatch = NULL;
+    if (dispatch == NULL)
     {
-        CV_LOG_DEBUG(NULL, "DNN/TF: processing TensorFlow config from memory (" << lenConfig << " bytes)");
-        ReadTFNetParamsFromTextBufferOrDie(dataConfig, lenConfig, &netTxt);
+        cv::AutoLock lock(getDispatchMapMutex());
+        if (dispatch == NULL)
+        {
+            dispatch = &getDispatchMapImpl();
```

Comment:
I might be wrong, but I think the compiler is free to rearrange `dispatch[layerName] = func` with `initialized = true` hence the Impl function. And `initialized` should be atomic, also the default constructor of `dispatch`can be ran by two threads at once. I think my solution is also not perfect since dispatch pointer isn't atomic either.

Reply:
AutoLock prevents malformed rearrangements.

Reply:
There is no guarantee `pDispatch` is not `NULL` when `initialized` is `true`, because memory may be partially out of sync across different threads. I'd like to force `pDispatch`  to update when `initialized` is read from, but it is only possible via release-acquire semantics of atomic's loads and stores. It means that `initialized`  has to be an equivalent of `atomic<bool>`, which we don't have. Another way is to lock the mutex prior to reading `initialized`.

Reply:
Ok, what is about to simplify C++11 code path?

```.diff
 const TFImporter::DispatchMap& TFImporter::getDispatchMap()
 {
+#ifdef CV_CXX11
+    static DispatchMap& dispatch = getDispatchMapImpl();
+    return dispatch;
+#else
     static const TFImporter::DispatchMap* volatile dispatch = NULL;
     if (dispatch == NULL)
     {
         cv::AutoLock lock(getDispatchMapMutex());
         if (dispatch == NULL)
         {
             dispatch = &getDispatchMapImpl();
         }
     }
     return *dispatch;
+#endif
 }

```